File: tests/e2e/lib/retry.ts
/**
 * Retry utilities for E2E tests
 *
 * Provides exponential backoff retry logic for operations that may take time
 * to propagate or may be temporarily unavailable.
 */

export interface RetryConfig {
  maxAttempts: number;
  initialDelayMs: number;
  maxDelayMs: number;
  backoffMultiplier: number;
}

/**
 * Default retry configuration for standard operations
 * 10 attempts, 100ms-5s delay range
 */
export function defaultRetryConfig(): RetryConfig {
  return {
    maxAttempts: 10,
    initialDelayMs: 100,
    maxDelayMs: 5000,
    backoffMultiplier: 2,
  };
}

/**
 * Quick retry configuration for fast operations
 * 5 attempts, 50ms-500ms delay range
 */
export function quickRetryConfig(): RetryConfig {
  return {
    maxAttempts: 5,
    initialDelayMs: 50,
    maxDelayMs: 500,
    backoffMultiplier: 2,
  };
}

/**
 * Slow retry configuration for external APIs or slow operations
 * 15 attempts, 500ms-10s delay range
 */
export function slowRetryConfig(): RetryConfig {
  return {
    maxAttempts: 15,
    initialDelayMs: 500,
    maxDelayMs: 10000,
    backoffMultiplier: 1.5,
  };
}

/**
 * Retry a function until it returns a truthy result or max attempts reached.
 *
 * @param fn - Async function to retry
 * @param config - Retry configuration
 * @returns The result of the function when it succeeds
 * @throws Error if max attempts exceeded
 *
 * @example
 * ```ts
 * const transaction = await retryUntil(
 *   async () => {
 *     const result = await api.getTransaction(id);
 *     return result.status === 'APPROVED' ? result : null;
 *   },
 *   defaultRetryConfig()
 * );
 * ```
 */
export async function retryUntil<T>(
  fn: () => Promise<T | null | undefined>,
  config: RetryConfig = defaultRetryConfig()
): Promise<T> {
  let delay = config.initialDelayMs;
  let lastError: Error | undefined;

  for (let attempt = 1; attempt <= config.maxAttempts; attempt++) {
    try {
      const result = await fn();
      if (result) {
        return result;
      }
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));
    }

    if (attempt < config.maxAttempts) {
      await sleep(delay);
      delay = Math.min(delay * config.backoffMultiplier, config.maxDelayMs);
    }
  }

  throw lastError || new Error(`Retry exhausted after ${config.maxAttempts} attempts`);
}

/**
 * Retry a function until a condition is met.
 *
 * @param fn - Async function to retry
 * @param condition - Condition function that returns true when done
 * @param config - Retry configuration
 * @returns The result of the function when condition is met
 *
 * @example
 * ```ts
 * const transaction = await retryUntilCondition(
 *   () => api.getTransaction(id),
 *   (result) => result.status === 'APPROVED',
 *   defaultRetryConfig()
 * );
 * ```
 */
export async function retryUntilCondition<T>(
  fn: () => Promise<T>,
  condition: (result: T) => boolean,
  config: RetryConfig = defaultRetryConfig()
): Promise<T> {
  let delay = config.initialDelayMs;
  let lastResult: T | undefined;
  let lastError: Error | undefined;

  for (let attempt = 1; attempt <= config.maxAttempts; attempt++) {
    try {
      lastResult = await fn();
      if (condition(lastResult)) {
        return lastResult;
      }
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));
    }

    if (attempt < config.maxAttempts) {
      await sleep(delay);
      delay = Math.min(delay * config.backoffMultiplier, config.maxDelayMs);
    }
  }

  if (lastError) {
    throw lastError;
  }
  if (lastResult !== undefined) {
    return lastResult; // Return last result even if condition wasn't met
  }
  throw new Error(`Retry exhausted after ${config.maxAttempts} attempts`);
}

/**
 * Wait for a condition to become true within a timeout.
 *
 * @param condition - Async condition function
 * @param timeoutMs - Maximum time to wait
 * @param intervalMs - Check interval
 * @returns true if condition was met, false if timeout
 *
 * @example
 * ```ts
 * const ready = await waitForCondition(
 *   async () => {
 *     const resp = await fetch(healthUrl);
 *     return resp.ok;
 *   },
 *   30000, // 30 second timeout
 *   1000   // check every second
 * );
 * ```
 */
export async function waitForCondition(
  condition: () => Promise<boolean>,
  timeoutMs: number,
  intervalMs: number = 1000
): Promise<boolean> {
  const startTime = Date.now();

  while (Date.now() - startTime < timeoutMs) {
    try {
      if (await condition()) {
        return true;
      }
    } catch {
      // Ignore errors and keep trying
    }
    await sleep(intervalMs);
  }

  return false;
}

/**
 * Wait for a server health endpoint to respond.
 *
 * @param healthUrl - URL to the health endpoint
 * @param timeoutMs - Maximum time to wait (default: 30s)
 * @returns true if server is ready, false if timeout
 *
 * @example
 * ```ts
 * const ready = await waitForServer('http://localhost:8081/cron/health');
 * if (!ready) {
 *   throw new Error('Server did not start in time');
 * }
 * ```
 */
export async function waitForServer(
  healthUrl: string,
  timeoutMs: number = 30000
): Promise<boolean> {
  return waitForCondition(
    async () => {
      try {
        const response = await fetch(healthUrl);
        return response.ok;
      } catch {
        return false;
      }
    },
    timeoutMs,
    1000
  );
}

/**
 * Sleep for a specified duration.
 */
function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

File: tests/integration/testutil/retry.go
package testutil

import (
	"context"
	"errors"
	"fmt"
	"testing"
	"time"
)

// RetryConfig configures retry behavior
type RetryConfig struct {
	MaxAttempts int           // Maximum number of attempts (default: 10)
	InitialWait time.Duration // Initial wait between attempts (default: 100ms)
	MaxWait     time.Duration // Maximum wait between attempts (default: 5s)
	Multiplier  float64       // Wait time multiplier (default: 2.0)
}

// DefaultRetryConfig returns sensible defaults for integration tests
func DefaultRetryConfig() RetryConfig {
	return RetryConfig{
		MaxAttempts: 10,
		InitialWait: 100 * time.Millisecond,
		MaxWait:     5 * time.Second,
		Multiplier:  2.0,
	}
}

// QuickRetryConfig returns config for fast-responding operations
func QuickRetryConfig() RetryConfig {
	return RetryConfig{
		MaxAttempts: 5,
		InitialWait: 50 * time.Millisecond,
		MaxWait:     500 * time.Millisecond,
		Multiplier:  1.5,
	}
}

// SlowRetryConfig returns config for slow operations (external APIs, DB heavy)
func SlowRetryConfig() RetryConfig {
	return RetryConfig{
		MaxAttempts: 15,
		InitialWait: 500 * time.Millisecond,
		MaxWait:     10 * time.Second,
		Multiplier:  2.0,
	}
}

// ErrMaxAttemptsReached is returned when all retry attempts are exhausted
var ErrMaxAttemptsReached = errors.New("max retry attempts reached")

// RetryUntil retries the given function until it returns true or max attempts reached
// Use this instead of time.Sleep() for polling operations
//
// Example:
//
//	err := testutil.RetryUntil(t, "wait for transaction", testutil.DefaultRetryConfig(), func() (bool, error) {
//	    tx, err := client.GetTransaction(txID)
//	    if err != nil {
//	        return false, nil // Keep retrying
//	    }
//	    return tx.Status == "approved", nil
//	})
func RetryUntil(t *testing.T, description string, cfg RetryConfig, fn func() (done bool, err error)) error {
	t.Helper()

	if cfg.MaxAttempts == 0 {
		cfg = DefaultRetryConfig()
	}

	wait := cfg.InitialWait
	var lastErr error

	for attempt := 1; attempt <= cfg.MaxAttempts; attempt++ {
		done, err := fn()
		if err != nil {
			lastErr = err
			t.Logf("%s: attempt %d/%d failed: %v", description, attempt, cfg.MaxAttempts, err)
		} else if done {
			if attempt > 1 {
				t.Logf("%s: succeeded on attempt %d", description, attempt)
			}
			return nil
		}

		if attempt < cfg.MaxAttempts {
			time.Sleep(wait)
			wait = time.Duration(float64(wait) * cfg.Multiplier)
			if wait > cfg.MaxWait {
				wait = cfg.MaxWait
			}
		}
	}

	if lastErr != nil {
		return fmt.Errorf("%s: %w (last error: %v)", description, ErrMaxAttemptsReached, lastErr)
	}
	return fmt.Errorf("%s: %w", description, ErrMaxAttemptsReached)
}

// RetryUntilWithContext is like RetryUntil but respects context cancellation
func RetryUntilWithContext(ctx context.Context, t *testing.T, description string, cfg RetryConfig, fn func() (done bool, err error)) error {
	t.Helper()

	if cfg.MaxAttempts == 0 {
		cfg = DefaultRetryConfig()
	}

	wait := cfg.InitialWait
	var lastErr error

	for attempt := 1; attempt <= cfg.MaxAttempts; attempt++ {
		select {
		case <-ctx.Done():
			return fmt.Errorf("%s: %w", description, ctx.Err())
		default:
		}

		done, err := fn()
		if err != nil {
			lastErr = err
			t.Logf("%s: attempt %d/%d failed: %v", description, attempt, cfg.MaxAttempts, err)
		} else if done {
			if attempt > 1 {
				t.Logf("%s: succeeded on attempt %d", description, attempt)
			}
			return nil
		}

		if attempt < cfg.MaxAttempts {
			select {
			case <-ctx.Done():
				return fmt.Errorf("%s: %w", description, ctx.Err())
			case <-time.After(wait):
			}
			wait = time.Duration(float64(wait) * cfg.Multiplier)
			if wait > cfg.MaxWait {
				wait = cfg.MaxWait
			}
		}
	}

	if lastErr != nil {
		return fmt.Errorf("%s: %w (last error: %v)", description, ErrMaxAttemptsReached, lastErr)
	}
	return fmt.Errorf("%s: %w", description, ErrMaxAttemptsReached)
}

// WaitForCondition is a simpler version of RetryUntil that just waits for a condition
// Returns error if condition is never met within timeout
//
// Example:
//
//	err := testutil.WaitForCondition(t, "server ready", 10*time.Second, func() bool {
//	    resp, err := http.Get(serverURL + "/health")
//	    return err == nil && resp.StatusCode == 200
//	})
func WaitForCondition(t *testing.T, description string, timeout time.Duration, condition func() bool) error {
	t.Helper()

	deadline := time.Now().Add(timeout)
	interval := 100 * time.Millisecond

	for time.Now().Before(deadline) {
		if condition() {
			return nil
		}
		time.Sleep(interval)
		// Exponential backoff up to 1 second
		if interval < time.Second {
			interval = time.Duration(float64(interval) * 1.5)
		}
	}

	return fmt.Errorf("%s: condition not met within %v", description, timeout)
}

// WaitForServer waits for an HTTP server to be ready (returns 200 on health endpoint)
func WaitForServer(t *testing.T, healthURL string, timeout time.Duration) error {
	t.Helper()

	return WaitForCondition(t, "server ready at "+healthURL, timeout, func() bool {
		client := NewClient("")
		resp, err := client.HTTPClient.Get(healthURL)
		if err != nil {
			return false
		}
		defer resp.Body.Close()
		return resp.StatusCode == 200
	})
}

// WaitForDatabase waits for a database to be ready
func WaitForDatabase(t *testing.T, timeout time.Duration) error {
	t.Helper()

	return WaitForCondition(t, "database ready", timeout, func() bool {
		db := GetDB(t)
		if db == nil {
			return false
		}
		return db.Ping() == nil
	})
}

// MustWait is a helper that fails the test if waiting fails
// Use this when the wait is critical and failure should stop the test
//
// Example:
//
//	testutil.MustWait(t, testutil.WaitForServer(t, cfg.ServiceURL+"/health", 30*time.Second))
func MustWait(t *testing.T, err error) {
	t.Helper()
	if err != nil {
		t.Fatalf("Wait failed: %v", err)
	}
}
